### PR TITLE
Add readonly type - Phase I

### PR DESCRIPTION
**Pull request by MaryamZi**
## Purpose
$title. 

Initial changes to fix https://github.com/ballerina-platform/ballerina-lang/issues/22430
Adds support for readonly record fields in the typedesc, partial fix for https://github.com/ballerina-platform/ballerina-lang/issues/22600

Fixes https://github.com/ballerina-platform/ballerina-lang/issues/22601

Working on improvements, adding tests, etc. Sending this PR since several tasks are blocked on this.

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java [hunk 1]
```diff
@@ -1054,22 +1054,107 @@ private boolean validateSpecifiedFields(BLangRecordLiteral mappingConstructor, B
...
-            // Check if `field` is explicitly assigned a value in the record literal
-            // If a required field is missing, it's a compile error
-            if (!specFieldNames.contains(field.name.value) && Symbols.isFlagOn(field.symbol.flags, Flags.REQUIRED)) {
+            String fieldName = field.name.value;
+            BType fieldType = symTable.semanticError;
+            DiagnosticPos fieldPos;
+
+            if (specFieldNames.contains(fieldName)) {
+                RecordLiteralNode.RecordField recordField = getRecordFieldByName(specifiedFields, fieldName);
+                if (recordField == null) {
```
Comment (pubudu91) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:1071]: Can this be `null` here, if the mapping constructor contains the field?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java [hunk 2]
```diff
@@ -1054,22 +1054,107 @@ private boolean validateSpecifiedFields(BLangRecordLiteral mappingConstructor, B
...
+                        BLangRecordLiteral.BLangRecordSpreadOperatorField spreadOpField =
+                                (BLangRecordLiteral.BLangRecordSpreadOperatorField) recordField;
+                        BRecordType recordType = (BRecordType) spreadOpField.expr.type;
+
+                        for (BField bField : recordType.fields) {
+                            if (fieldName.equals(bField.name.value)) {
+                                fieldType = bField.type;
+                                break;
+                            }
+                        }
```
Comment (pubudu91) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:1097]: Isn't this part duplicate of the logic in `getRecordFieldByName()`? Do we need to do the same here, after calling that method? Should be able to directly take the type of the spread field right?

File: compiler/ballerina-lang/src/main/resources/compiler.properties [hunk 3]
```diff
@@ -347,6 +347,9 @@ error.invalid.record.rest.descriptor=\
 error.missing.required.record.field=\
   missing non-defaultable required record field ''{0}''
 
+error.invalid.field.for.readonly.record.field=\
+  invalid field; ''readonly'' field expected for ''{0}''
```
Comment (pubudu91) [inline compiler/ballerina-lang/src/main/resources/compiler.properties:351]: Shouldn't this be a colon?

Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/resources/compiler.properties]: Fixed, but we seem to have used a mix of `:` and `;` in error messages.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java (hunk 1 above)
Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java]: Yeah not in this context, I've removed the null check.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java (hunk 2 above)
Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java]: Yeah, noticed the duplication, but the issue was that we basically needed the type and pos, and for key-value and variable name fields we can retrieve a `RecordField` and get the type and pos from it. But for a spread-field the type is from a `BField` and the pos is still from the spread field. 

I've refactored this a bit with an inner class.

File: tests/jballerina-unit-test/src/test/resources/test-src/record/readonly_record_fields.bal [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+};
+
+function testValidUpdateOfPossiblyReadonlyFieldInUnion() {
+    Details d = {
+        name: "Jo",
+        id: 1234
+    };
+
+    Student|Details sd = d;
+    sd.name = "May";
```
Comment (hasithaa) [inline tests/jballerina-unit-test/src/test/resources/test-src/record/readonly_record_fields.bal:94]: Need to clarify this behavior. `Student.name` is a readonly field, So `sd.name` assignment can panic at runtime. 
We have similar behavior on the inherent type violation. So this is OK IMO.

Comment (MaryamZi) [reply tests/jballerina-unit-test/src/test/resources/test-src/record/readonly_record_fields.bal:94]: Created https://github.com/ballerina-platform/ballerina-spec/issues/495 to clarify.

**Pull request merged**